apps/openmw/mwclass/container.cpp

```diff
@@ -20,57 +20,186 @@
 #include "../mwworld/actiontrap.hpp"
 #include "../mwphysics/physicssystem.hpp"
 #include "../mwworld/inventorystore.hpp"
+#include "../mwworld/localscripts.hpp"
 
 #include "../mwgui/tooltips.hpp"
 
 #include "../mwrender/animation.hpp"
 #include "../mwrender/objects.hpp"
 #include "../mwrender/renderinginterface.hpp"
 
+#include "../mwmechanics/actorutil.hpp"
 #include "../mwmechanics/npcstats.hpp"
 
+namespace
+{
+    void removeScripts(const MWWorld::ContainerStore& store)
+    {
+        auto& scripts = MWBase::Environment::get().getWorld()->getLocalScripts();
+        for(const MWWorld::ConstPtr& ptr : store)
+        {
+            if(!ptr.getClass().getScript(ptr).empty())
+               scripts.remove(ptr);
+        }
+    }
+
+    void addScripts(MWWorld::ContainerStore& store, MWWorld::CellStore* cell)
+    {
+        auto& scripts = MWBase::Environment::get().getWorld()->getLocalScripts();
+        for(const MWWorld::Ptr& ptr : store)
+        {
+            const std::string& script = ptr.getClass().getScript(ptr);
+            if(!script.empty())
+            {
+                MWWorld::Ptr item = ptr;
+                item.mCell = cell;
+                scripts.add(script, item);
+            }
+        }
+    }
+}
+
 namespace MWClass
 {
-    class ContainerCustomData : public MWWorld::CustomData
+    class ResolutionListener
     {
+        ContainerCustomData& mCustomData;
+        MWWorld::CellStore* mCell;
     public:
-        MWWorld::ContainerStore mContainerStore;
+        ResolutionListener(ContainerCustomData& customData, MWWorld::CellStore* cell) : mCustomData(customData), mCell(cell) {}
+
+        ~ResolutionListener();
+    };
+
+    class ResolvingStoreManager : public MWWorld::ContainerStoreProvider
+    {
+        ContainerCustomData& mCustomData;
+        const ESM::Container& mContainer;
+        MWWorld::CellStore* mCell;
+        mutable std::shared_ptr<ResolutionListener> mListener;
+    public:
+        ResolvingStoreManager(ContainerCustomData& customData, const ESM::Container& container, MWWorld::CellStore* cell)
+        : mCustomData(customData), mContainer(container), mCell(cell) {}
+
+        virtual MWWorld::ContainerStore& getMutable() override;
+
+        virtual const MWWorld::ContainerStore& getImmutable() const override;
+    };
 
-        virtual MWWorld::CustomData *clone() const;
+    ContainerCustomData::ContainerCustomData(const ESM::Container& container, MWWorld::CellStore* cell)
+    : mUnresolvedStore(std::make_unique<MWWorld::ContainerStore>())
+    , mSeed(Misc::Rng::rollDice(std::numeric_limits<int>::max())) {
```

Comment:
It is unclear why you need a separate seed and separate random generator just for every container. IMO, it is an overkill.

Reply:
The seed is so it resolves to the same contents every time it's opened - unless the player's level changed. Which is vanilla behaviour.

I figured I'd create Rng instances instead of changing the global number generator's seed so the player doesn't end up messing with random number generation by opening containers.

Reply:
> The seed is so it resolves to the same contents every time it's opened

I suppose you can just resolve items once, but do not store them in savegames. Approach with separate random generator for every container looks like a hack.

Reply:
That doesn't work because the contents can still change. Levelled list resolution is level-dependent (obviously), if the player opens a container it will always contain the same items provided the game hasn't been restarted and the player's level hasn't changed.

Reply:
> Levelled list resolution is level-dependent (obviously)

And? You can store current container level as a temporary field and re-generate content on opening when player's level is changed.

It is still better than re-generate the whole content with its scripts on every access.

Reply:
As it stands resolution is a rare event so that sounds more like premature optimization, but yeah it could work.